### PR TITLE
fix: upgrade user id strategy to string types rather than numerics

### PR DESCRIPTION
**Pull request by sighphyre**
User ids can be string types, prior to this, Yggdrasil was trying to upgrade these to numeric rules like so:

`user_in IN [1,2,3]`

This doesn't work when the user supplies a user id which doesn't parse as a numeric. This PR changes this so that the strategy is upgraded to a string type like so:

`user_in IN ["1","2","3"]`

### CODE REVIEW

File: unleash-yggdrasil/src/strategy_parsing.rs [hunk 1]
```diff
@@ -689,6 +689,8 @@ mod tests {
     #[test_case("user_id in [1, 3, 6]", true)]
     #[test_case("user_id in [1, 3, 5]", false)]
     #[test_case("user_id not_in [1, 3, 5]", true)]
+    #[test_case("user_id in [\"dfsfsd\"]", false)]
```
Comment (sighphyre) [inline unleash-yggdrasil/src/strategy_parsing.rs:692]: The grammar supports this case out the box but it's good to layer a test over this so we know for future

File: unleash-yggdrasil/src/strategy_parsing.rs [hunk 2]
```diff
@@ -689,6 +689,8 @@ mod tests {
...
+    #[test_case("user_id not_in [\"dfsfsd\"]", true)]
```
Comment (gastonfournier) [inline unleash-yggdrasil/src/strategy_parsing.rs:693]: Just a minor detail: with this is not clear why `user_id not_in [\"dfsfsd\"]` should be true without looking at the body of the test. I don't know if it's worth adding the `6` parameter from where you're taking the context. 

Something like:
1. scenario: given the context for user id 6
2. when: user_id in [\"dfsfsd\"]
3. then: should be false

**Review by gastonfournier — APPROVED** [review]
Looks fine! Just a small detail that you can ignore (all other tests in this file works similarly)

Also, I noticed that Rust files tend to be really large due to embedded tests

File: unleash-yggdrasil/src/strategy_upgrade.rs [hunk 3]
```diff
@@ -99,7 +99,14 @@
 
 fn upgrade_user_id_strategy(strategy: &Strategy) -> String {
     match strategy.get_param("userIds") {
-        Some(user_ids) => format!("user_id in [{}]", user_ids),
+        Some(user_ids) => {
+            let user_ids = user_ids
+                .split(",")
```
[the remark this answers is not in the document]
Comment (sighphyre) [reply unleash-yggdrasil/src/strategy_upgrade.rs]: Fixed

File: unleash-yggdrasil/src/strategy_parsing.rs (hunk 2 above)
Comment (sighphyre) [reply unleash-yggdrasil/src/strategy_parsing.rs:693]: > Just a minor detail: with this is not clear why user_id not_in [\"dfsfsd\"] should be true without looking at the body of the test. I don't know if it's worth adding the 6 parameter from where you're taking the context.

Ah. Yeah fair enough. I've gotten too used to reading these. Probably not a thing I want to do in this PR but a big refactor of the test names might be worthwhile now that I'm no longer the only person poking

> Also, I noticed that Rust files tend to be really large due to embedded tests

Yeah, you get used to it over time. You start to look for the mod test and mark that in your mind as the relevant test code. Takes a bit of time but I prefer it now

Comment (gastonfournier) [reply unleash-yggdrasil/src/strategy_parsing.rs:693]: > Yeah, you get used to it over time. You start to look for the mod test and mark that in your mind as the relevant test code. Takes a bit of time but I prefer it now

Challenge accepted :rofl:

**Pull request merged**